src/student_success_tool/analysis/pdp/targets/failure_to_graduate_within_x_years.py

```diff
@@ -0,0 +1,208 @@
+        student_id_cols=student_id_cols,
+        intensity_time_lefts=intensity_time_lefts,
+        max_term_rank=max_term_rank,
+        num_terms_in_year=num_terms_in_year,
+        exclude_pre_cohort_terms=exclude_pre_cohort_terms,
+        term_is_pre_cohort_col=term_is_pre_cohort_col,
+        n=n,
+        enrollment_intensity_col=enrollment_intensity_col,
+        term_rank_col=term_rank_col,
+    )
+    df_eligible_student_terms = pd.merge(
+        df, df_eligible_students, on=student_id_cols, how="inner"
+    )
+    if exclude_pre_cohort_terms is True:
+        df_features = shared.get_nth_student_terms_within_cohort(
+            df_eligible_student_terms,
+            student_id_cols=student_id_cols,
+            term_is_pre_cohort_col=term_is_pre_cohort_col,
+            sort_cols=term_rank_col,
+            n=n,
+            include_cols=None,
+        )
+    else:
+        df_features = shared.get_nth_student_terms(
+            df_eligible_student_terms,
+            student_id_cols=student_id_cols,
+            sort_cols=term_rank_col,
+            n=n,
+            include_cols=None,
+        )
+    df_targets = compute_target_variable(
+        df_eligible_student_terms,
+        intensity_time_lefts=intensity_time_lefts,
+        student_id_cols=student_id_cols,
+        enrollment_intensity_col=enrollment_intensity_col,
+        years_to_degree_col=years_to_degree_col,
+        num_terms_in_year=num_terms_in_year,
+    )
+    df_labeled = pd.merge(df_features, df_targets, on=student_id_cols, how="inner")
+    return df_labeled
+
+
+def compute_target_variable(    
+    df: pd.DataFrame,    
+    *,    
+    student_id_cols: str | list[str] = "student_guid",    
+    years_to_degree_col: str = "first_year_bach",  
+    intensity_time_lefts: list[tuple[str, float, t.Literal["year", "term"]]],   
+    intensity_col: str = "intensity_type",  
+    enrollment_intensity_col: str = "enrollment_intensity_first_term",
+    num_terms_in_year: int = 4,
+) -> pd.Series:   
+    """  
+    Args:  
+        df  
+        student_id_cols  
+        years_to_degree_col  
+        intensity_time_lefts  
+        intensity_col  
+    """   
+    student_id_cols = utils.to_list(student_id_cols)  
+    intensity_num_terms = {intensity: time if unit == "year" else time / num_terms_in_year  
+                           for intensity, time, unit in intensity_time_lefts  
+                           }   
+    return (    
+            df[student_id_cols + [years_to_degree_col, enrollment_intensity_col]]    
+            .drop_duplicates(subset=student_id_cols, ignore_index=True)    
+            .assign(target=lambda df: (df[years_to_degree_col].gt(df[enrollment_intensity_col].map(intensity_num_terms))) |   
+                                     (df[years_to_degree_col].isna()))       
+            .astype({"target": "bool"})    
+            .set_index(student_id_cols)    
+            .loc[:, "target"]    
+            )    
```

Comment:
Could you format this code using `ruff format` ? It's hard to read with the whitespace as it is 🙏

Reply:
How do i run this in terminal in databricks?

Reply:
nevermind figured it out!

Reply:
You were able to run this in the terminal in databricks? Or you realized that this is the public repo, so you can run it locally via your text editor?

Reply:
in the terminal in databricks :)